test/util/TestFeatureRegistry.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <memory>
+
+#include "../../source/FeatureRegistry.h"
+#include "../../source/config/Config.h"
+#include "gtest/gtest.h"
+
+using namespace std;
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+
+class FakeFeature : public Feature
+{
+  public:
+    FakeFeature(const std::string &name) : name(name) {}
+    int start() override
+    {
+        started = true;
+        stopped = false;
+        return 0;
+    }
+    int stop() override
+    {
+        stopped = true;
+        started = false;
+        return 0;
+    }
+    std::string getName() { return name; }
+    bool isStarted() { return started; }
+    bool isStopped() { return stopped; }
+
+  private:
+    std::string name;
+    bool started;
+    bool stopped;
+};
+
+class TestFeatureRegistry : public ::testing::Test
+{
+  public:
+    shared_ptr<FeatureRegistry> features;
+    shared_ptr<FakeFeature> feature1;
+    shared_ptr<FakeFeature> feature2;
+    shared_ptr<FakeFeature> feature3;
+
+    void SetUp() override
+    {
+        features = make_shared<FeatureRegistry>();
+        feature1 = make_shared<FakeFeature>("feature-1");
+        feature2 = make_shared<FakeFeature>("feature-2");
+        feature3 = make_shared<FakeFeature>("feature-3");
+    }
+
+    void TearDown() override { features = make_shared<FeatureRegistry>(); }
```

Comment:
I don't think this is needed? Could be wrong though.

the features shared_ptr will be reinitialized before each test by the SetUp method.